File: packages/runtime/src/file.ts
import * as tg from "./index.ts";
import { unindent } from "./template.ts";

export async function file(
	strings: TemplateStringsArray,
	...placeholders: tg.Args<string>
): Promise<tg.File>;
export async function file(...args: tg.Args<tg.File.Arg>): Promise<tg.File>;
export async function file(
	firstArg:
		| TemplateStringsArray
		| tg.Unresolved<tg.ValueOrMaybeMutationMap<tg.File.Arg>>,
	...args: tg.Args<tg.File.Arg>
): Promise<tg.File> {
	return await inner(false, firstArg, ...args);
}

async function inner(
	raw: boolean,
	firstArg:
		| TemplateStringsArray
		| tg.Unresolved<tg.ValueOrMaybeMutationMap<tg.File.Arg>>,
	...args: tg.Args<tg.File.Arg>
): Promise<tg.File> {
	if (Array.isArray(firstArg) && "raw" in firstArg) {
		let strings = firstArg;
		let placeholders = args as tg.Args<string>;
		let components = [];
		for (let i = 0; i < strings.length - 1; i++) {
			let string = strings[i]!;
			components.push(string);
			let placeholder = placeholders[i]!;
			components.push(placeholder);
		}
		components.push(strings[strings.length - 1]!);
		let string = components.join("");
		if (!raw) {
			string = unindent([string]).join("");
		}
		return await File.new(string);
	} else {
		return await File.new(firstArg as tg.File.Arg, ...args);
	}
}

export class File {
	#state: File.State;

	constructor(state: File.State) {
		this.#state = state;
	}

	get state(): File.State {
		return this.#state;
	}

	static withId(id: File.Id): File {
		return new File({ id, stored: true });
	}

	static withObject(object: File.Object): File {
		return new File({ object, stored: false });
	}

	static fromData(data: File.Data): File {
		return File.withObject(File.Object.fromData(data));
	}

	static async new(...args: tg.Args<File.Arg>): Promise<File> {
		if (args.length === 1) {
			let arg = await tg.resolve(args[0]);
			if (typeof arg === "object" && "node" in arg) {
				return tg.File.withObject(arg as tg.Graph.Reference);
			}
		}
		let arg = await tg.File.arg(...args);
		let contents = await tg.blob(arg.contents);
		let dependencies = Object.fromEntries(
			Object.entries(arg.dependencies ?? {}).map(([key, value]) => {
				let referent: tg.Referent<tg.Graph.Edge<tg.Object>> | undefined;
				if (
					typeof value === "number" ||
					(value && "node" in value) ||
					tg.Object.is(value)
				) {
					let item = tg.Graph.Edge.fromArg(value);
					referent = { item, options: {} };
				} else if (value) {
					let item = tg.Graph.Edge.fromArg(value.item);
					referent = { item, options: value.options };
				}
				return [key, referent];
			}),
		);
		let executable = arg.executable ?? false;
		let object = { contents, dependencies, executable };
		return tg.File.withObject(object);
	}

	static async arg(
		...args: tg.Args<tg.File.Arg>
	): Promise<Exclude<tg.File.Arg.Object, tg.Graph.Arg.Reference>> {
		type Arg = Exclude<tg.File.Arg.Object, tg.Graph.Arg.Reference>;
		return await tg.Args.apply<tg.File.Arg, Arg>({
			args,
			map: async (arg) => {
				if (arg === undefined) {
					return {};
				} else if (
					typeof arg === "string" ||
					arg instanceof Uint8Array ||
					arg instanceof tg.Blob
				) {
					return { contents: arg };
				} else if (arg instanceof File) {
					return {
						contents: await arg.contents(),
						dependencies: await arg.dependencies(),
					};
				} else {
					return arg as Arg;
				}
			},
			reduce: {
				contents: (a, b) => tg.blob(a, b),
				dependencies: "merge",
			},
		});
	}

	static expect(value: unknown): tg.File {
		tg.assert(value instanceof File);
		return value;
	}

	static assert(value: unknown): asserts value is tg.File {
		tg.assert(value instanceof File);
	}

	get id(): tg.File.Id {
		if (this.#state.id! !== undefined) {
			return this.#state.id;
		}
		let object = this.#state.object!;
		let data = File.Object.toData(object);
		let id = syscall("object_id", { kind: "file", value: data });
		this.#state.id = id;
		return id;
	}

	async object(): Promise<File.Object> {
		await this.load();
		return this.#state.object!;
	}

	async load(): Promise<tg.File.Object> {
		if (this.#state.object === undefined) {
			let data = await syscall("object_get", this.#state.id!);
			tg.assert(data.kind === "file");
			let object = File.Object.fromData(data.value);
			this.#state.object = object;
		}
		return this.#state.object!;
	}

	async store(): Promise<tg.File.Id> {
		await tg.Value.store(this);
		return this.id;
	}

	async children(): Promise<Array<tg.Object>> {
		let object = await this.load();
		return tg.File.Object.children(object);
	}

	async contents(): Promise<tg.Blob> {
		let object = await this.object();
		if ("node" in object) {
			let graph = object.graph;
			tg.assert(graph !== undefined);
			let nodes = await graph.nodes();
			let node = nodes[object.node];
			tg.assert(node !== undefined);
			tg.assert(node.kind === "file");
			return node.contents;
		} else {
			tg.assert(object.contents);
			return object.contents;
		}
	}

	async dependencies(): Promise<{
		[reference: tg.Reference]: tg.Referent<tg.Object> | undefined;
	}> {
		let object = await this.object();
		if ("node" in object) {
			let graph = object.graph;
			tg.assert(graph !== undefined);
			let nodes = await graph.nodes();
			let node = nodes[object.node];
			tg.assert(node !== undefined);
			tg.assert(node.kind === "file");
			let dependencies = node.dependencies;
			return Object.fromEntries(
				await Promise.all(
					Object.entries(dependencies).map(async ([reference, referent]) => {
						if (!referent) {
							return [reference, undefined];
						}
						let object: tg.Object | undefined;
						if (typeof referent.item === "number") {
							object = await graph.get(referent.item);
						} else if ("node" in referent.item) {
							object = await (referent.item.graph ?? graph).get(
								referent.item.node,
							);
						} else {
							object = referent.item;
						}
						let value = {
							item: object,
							options: referent.options,
						};
						return [reference, value];
					}),
				),
			);
		} else {
			let dependencies = object.dependencies;
			return Object.fromEntries(
				await Promise.all(
					Object.entries(dependencies).map(async ([reference, referent]) => {
						if (!referent) {
							return [reference, undefined];
						}
						let object: tg.Object | undefined;
						tg.assert(typeof referent.item === "object");
						if ("node" in referent.item) {
							tg.assert(referent.item.graph !== undefined);
							object = await referent.item.graph.get(referent.item.node);
						} else {
							object = referent.item;
						}
						let value = {
							item: object,
							options: referent.options,
						};
						return [reference, value];
					}),
				),
			);
		}
	}

	async dependencyObjects(): Promise<Array<tg.Object>> {
		let dependencies = await this.dependencies();
		if (dependencies === undefined) {
			return [];
		} else {
			let items = [];
			for (let dependency of Object.values(dependencies)) {
				if (dependency) {
					items.push(dependency.item);
				}
			}
			return items;
		}
	}

	async executable(): Promise<boolean> {
		let object = await this.object();
		if ("node" in object) {
			let graph = object.graph;
			tg.assert(graph !== undefined);
			let nodes = await graph.nodes();
			let node = nodes[object.node];
			tg.assert(node !== undefined);
			tg.assert(node.kind === "file");
			return node.executable;
		} else {
			return object.executable;
		}
	}

	async length(): Promise<number> {
		return (await this.contents()).length();
	}

	async read(arg?: tg.Blob.ReadArg): Promise<Uint8Array> {
		return (await this.contents()).read(arg);
	}

	async bytes(): Promise<Uint8Array> {
		return (await this.contents()).bytes();
	}

	async text(): Promise<string> {
		return (await this.contents()).text();
	}
}

export namespace File {
	export type Id = string;

	export type State = tg.Object.State<File.Id, File.Object>;

	export type Arg =
		| undefined
		| string
		| Uint8Array
		| tg.Blob
		| tg.File
		| tg.File.Arg.Object;

	export namespace Arg {
		export type Object = tg.Graph.Arg.Reference | tg.Graph.Arg.File;
	}

	export type Object = tg.Graph.Reference | tg.Graph.File;

	export namespace Object {
		export let toData = (object: tg.File.Object): tg.File.Data => {
			if ("node" in object) {
				return tg.Graph.Reference.toData(object);
			} else {
				return tg.Graph.File.toData(object);
			}
		};

		export let fromData = (data: tg.File.Data): tg.File.Object => {
			if (tg.Graph.Data.Reference.is(data)) {
				return tg.Graph.Reference.fromData(data);
			} else {
				return tg.Graph.File.fromData(data);
			}
		};

		export let children = (object: tg.File.Object): Array<tg.Object> => {
			if ("node" in object) {
				return tg.Graph.Reference.children(object);
			} else {
				return tg.Graph.File.children(object);
			}
		};
	}

	export type Data = tg.Graph.Data.Reference | tg.Graph.Data.File;

	export let raw = async (
		strings: TemplateStringsArray,
		...placeholders: tg.Args<string>
	): Promise<File> => {
		return await inner(true, strings, ...placeholders);
	};
}

File: packages/clients/python/tangram/file.py
"""File class for Tangram.

This module mirrors the JavaScript packages/clients/js/src/file.ts.
"""

from __future__ import annotations

import base64
from typing import Any, Dict, List, Optional, Union

from tangram.object_ import Object_


async def file(*args: Any) -> "File":
    """Create a new file."""
    return await File.new(*args)


class File:
    """A file is an artifact with contents, dependencies, and executable flag."""

    _state: Object_.State

    def __init__(
        self,
        *,
        id: Optional[str] = None,
        object: Optional[Dict[str, Any]] = None,
        stored: bool = False,
    ) -> None:
        """Initialize a file."""
        obj = {"kind": "file", "value": object} if object is not None else None
        self._state = Object_.State(id=id, object=obj, stored=stored)

    @property
    def state(self) -> Object_.State:
        """Get the object state."""
        return self._state

    @classmethod
    def with_id(cls, id: str) -> "File":
        """Create a file with the given ID."""
        return cls(id=id, stored=True)

    @classmethod
    def with_object(cls, object: Dict[str, Any]) -> "File":
        """Create a file with the given object."""
        return cls(object=object, stored=False)

    @classmethod
    def from_data(cls, data: Dict[str, Any]) -> "File":
        """Create a file from serialized data."""
        return cls.with_object(File.Object.from_data(data))

    @classmethod
    async def new(cls, *args: Any) -> "File":
        """Create a new file from arguments."""
        import tangram as tg

        arg = await cls.arg(*args)
        contents = await tg.blob(arg.get("contents"))
        dependencies = {}
        for reference, value in (arg.get("dependencies") or {}).items():
            dependency: Optional[Dict[str, Any]] = None
            if value is None:
                dependency = None
            elif isinstance(value, int) or (isinstance(value, dict) and "index" in value) or Object_.is_(value):
                item = value
                dependency = {"item": item, "options": {}}
            elif isinstance(value, dict):
                item = value.get("item")
                dependency = {"item": item, "options": value.get("options", {})}
            dependencies[reference] = dependency
        executable = arg.get("executable", False)
        object = {"contents": contents, "dependencies": dependencies, "executable": executable}
        return cls.with_object(object)

    @classmethod
    async def arg(cls, *args: Any) -> Dict[str, Any]:
        """Process file arguments using Args.apply."""
        import tangram as tg

        async def map_fn(arg: Any) -> Dict[str, Any]:
            if arg is None:
                return {}
            elif isinstance(arg, (str, bytes)):
                return {"contents": arg}
            elif isinstance(arg, tg.Blob):
                return {"contents": arg}
            elif isinstance(arg, File):
                return {
                    "contents": await arg.contents(),
                    "dependencies": await arg.dependencies(),
                }
            elif isinstance(arg, dict):
                return arg
            else:
                return {}

        async def contents_reducer(a: Any, b: Any) -> Any:
            return await tg.blob(a, b)

        return await tg.Args.apply(
            args=list(args),
            map_fn=map_fn,
            reduce_spec={
                "contents": contents_reducer,
                "dependencies": "merge",
            },
        )

    @staticmethod
    def expect(value: Any) -> "File":
        """Assert that value is a File and return it."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, File))
        return value

    @staticmethod
    def assert_(value: Any) -> None:
        """Assert that value is a File."""
        from tangram.assert_ import assert_

        assert_(isinstance(value, File))

    @property
    def id(self) -> str:
        """Get the file ID."""
        from tangram.assert_ import assert_

        id = self._state.id
        assert_(Object_.Id.kind(id) == "file")
        return id

    async def object(self) -> Dict[str, Any]:
        """Get the file object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "file")
        return obj.get("value", obj)

    async def load(self) -> Dict[str, Any]:
        """Load the file object."""
        from tangram.assert_ import assert_

        obj = await self._state.load()
        assert_(obj.get("kind") == "file")
        return obj.get("value", obj)

    def unload(self) -> None:
        """Unload the file data."""
        self._state.unload()

    async def store(self) -> str:
        """Store the file and return its ID."""
        import tangram as tg

        await tg.Value.store(self)
        return self.id

    async def children(self) -> List[Object_]:
        """Get child objects."""
        return await self._state.children()

    async def contents(self) -> "tg.Blob":
        """Get the file contents as a blob."""
        import tangram as tg

        obj = await self.object()
        contents = obj.get("contents")
        if contents is None:
            return await tg.blob(b"")
        if isinstance(contents, tg.Blob):
            return contents
        return tg.Blob.with_id(contents)

    async def dependencies(self) -> Dict[str, Any]:
        """Get the file dependencies."""
        obj = await self.object()
        dependencies = obj.get("dependencies", {})
        result = {}
        for reference, dependency in dependencies.items():
            if dependency is None:
                result[reference] = None
            else:
                item = dependency.get("item")
                options = dependency.get("options", {})
                obj_item: Optional[Object_] = None
                if item is None:
                    obj_item = None
                elif isinstance(item, Object_):
                    obj_item = item
                elif isinstance(item, str):
                    obj_item = Object_.with_id(item)
                result[reference] = {"item": obj_item, "options": options}
        return result

    async def dependency_objects(self) -> List[Object_]:
        """Get all dependency objects."""
        dependencies = await self.dependencies()
        items = []
        for dependency in dependencies.values():
            if dependency is not None and dependency.get("item") is not None:
                items.append(dependency["item"])
        return items

    async def executable(self) -> bool:
        """Check if the file is executable."""
        obj = await self.object()
        return obj.get("executable", False)

    async def length(self) -> int:
        """Get the length of the file contents."""
        contents = await self.contents()
        return await contents.length()

    async def read(
        self,
        position: Optional[int] = None,
        length: Optional[int] = None,
        size: Optional[int] = None,
    ) -> bytes:
        """Read file contents."""
        contents = await self.contents()
        return await contents.read(position=position, length=length, size=size)

    async def bytes(self) -> bytes:
        """Get the file contents as bytes."""
        contents = await self.contents()
        return await contents.bytes()

    async def text(self) -> str:
        """Get the file contents as text."""
        contents = await self.contents()
        return await contents.text()

    # Nested Object namespace.
    class Object:
        """Namespace for File.Object operations."""

        @staticmethod
        def to_data(obj: Dict[str, Any]) -> Dict[str, Any]:
            """Convert object to data."""
            from tangram.blob import Blob

            data: Dict[str, Any] = {}
            contents = obj.get("contents")
            if contents is not None:
                if isinstance(contents, Blob):
                    data["contents"] = contents.id
                else:
                    data["contents"] = contents
            dependencies = obj.get("dependencies", {})
            if dependencies:
                data["dependencies"] = {}
                for reference, dependency in dependencies.items():
                    if dependency is None:
                        data["dependencies"][reference] = None
                    else:
                        item = dependency.get("item")
                        if item is None:
                            item_data = None
                        elif hasattr(item, "id"):
                            item_data = item.id
                        else:
                            item_data = item
                        data["dependencies"][reference] = {
                            "item": item_data,
                            "options": dependency.get("options", {}),
                        }
            executable = obj.get("executable", False)
            if executable:
                data["executable"] = executable
            return data

        @staticmethod
        def from_data(data: Dict[str, Any]) -> Dict[str, Any]:
            """Convert data to object."""
            from tangram.blob import Blob

            obj: Dict[str, Any] = {}
            contents = data.get("contents")
            if contents is not None:
                obj["contents"] = Blob.with_id(contents)
            dependencies_data = data.get("dependencies", {})
            dependencies = {}
            for reference, dependency_data in dependencies_data.items():
                if dependency_data is None:
                    dependencies[reference] = None
                else:
                    item_data = dependency_data.get("item")
                    if item_data is None:
                        item = None
                    elif isinstance(item_data, str):
                        item = Object_.with_id(item_data)
                    else:
                        item = item_data
                    dependencies[reference] = {
                        "item": item,
                        "options": dependency_data.get("options", {}),
                    }
            obj["dependencies"] = dependencies
            obj["executable"] = data.get("executable", False)
            return obj

        @staticmethod
        def children(obj: Optional[Dict[str, Any]]) -> List[Object_]:
            """Get child objects from file object."""
            from tangram.blob import Blob

            if obj is None:
                return []
            result: List[Object_] = []
            contents = obj.get("contents")
            if contents is not None:
                if isinstance(contents, Blob):
                    result.append(contents)
            dependencies = obj.get("dependencies", {})
            for dependency in dependencies.values():
                if dependency is not None:
                    item = dependency.get("item")
                    if item is not None and isinstance(item, Object_):
                        result.append(item)
            return result


# Type aliases.
File.Id = str
File.Arg = Union[None, str, bytes, "tg.Blob", "File", Dict[str, Any]]
File.Data = Dict[str, Any]
